src/ucp/proto/proto_am.inl

```diff
@@ -43,53 +50,52 @@ ucp_do_am_bcopy_single(uct_pending_req_t *self, uint8_t am_id,
                                    uint8_t am_id_middle,
-                                   size_t hdr_size_middle,
                                    uct_pack_callback_t pack_first,
                                    uct_pack_callback_t pack_middle,
                                    int enable_am_bw)
 {
-    ucp_request_t *req = ucs_container_of(self, ucp_request_t, send.uct);
-    ucp_ep_t *ep       = req->send.ep;
+    ucp_request_t *req   = ucs_container_of(self, ucp_request_t, send.uct);
+    ucp_ep_t *ep         = req->send.ep;
+    ucp_dt_state_t state = req->send.state.dt;
     ucs_status_t status;
-    size_t UCS_V_UNUSED max_middle;
     ssize_t packed_len;
     uct_ep_h uct_ep;
-    size_t offset;
     int pending_adde_res;
 
-    offset         = req->send.state.dt.offset;
-    req->send.lane = (!enable_am_bw || !offset) ? /* first part of message must be sent */
-                     ucp_ep_get_am_lane(ep) :     /* via AM lane */
+    req->send.lane = (!enable_am_bw || (state.offset == 0)) ? /* first part of message must be sent */
+                     ucp_ep_get_am_lane(ep) :                 /* via AM lane */
                      ucp_send_request_get_next_am_bw_lane(req);
     uct_ep         = ep->uct_eps[req->send.lane];
-    max_middle     = ucp_ep_get_max_bcopy(ep, req->send.lane) - hdr_size_middle;
 
     for (;;) {
-        if (offset == 0) {
+        if (state.offset == 0) {
             /* First */
             packed_len = uct_ep_am_bcopy(uct_ep, am_id_first, pack_first, req, 0);
-            UCS_PROFILE_REQUEST_EVENT_CHECK_STATUS(req, "am_bcopy_first", packed_len,
-                                                   packed_len);
-            ucs_assertv(req->send.state.dt.offset < req->send.length,
-                        "offset=%zd", req->send.state.dt.offset);
+            UCS_PROFILE_REQUEST_EVENT_CHECK_STATUS(req, "am_bcopy_first",
+                                                   packed_len, packed_len);
         } else {
-            ucs_assert(offset < req->send.length);
+            ucs_assert(state.offset < req->send.length);
             /* Middle or last */
             packed_len = uct_ep_am_bcopy(uct_ep, am_id_middle, pack_middle, req, 0);
-            ucs_assertv((packed_len < 0) || (packed_len <= max_middle + hdr_size_middle),
-                        "packed_len=%zd max_middle=%zu hdr_size_middle=%zu",
-                        packed_len, max_middle, hdr_size_middle);
             UCS_PROFILE_REQUEST_EVENT_CHECK_STATUS(req, "am_bcopy_middle",
                                                    packed_len, packed_len);
-            ucs_assert((packed_len < 0) ||
-                       (offset + packed_len - hdr_size_middle <= req->send.length));
-            if ((packed_len > 0) && (offset + packed_len - hdr_size_middle == req->send.length)) {
-                /* Last */
-                return UCS_OK;
-            }
         }
 
+        ucs_assertv((packed_len < 0) ||
```

Comment:
i'd removed (`packed_len < 0`) and move the assert under `else` on line 112

Reply:
good point, will do